File: Yr2022/day_eleven.py
from time import perf_counter_ns
from dataclasses import dataclass
from typing import Deque, Union
from collections import deque
from tqdm import tqdm


def cmd_eval(old, eval_str):
    return eval(eval_str)


@dataclass
class Item:
    init_val: int
    def setup(self, divisor_list):
        self.remainders = {k : self.init_val % k for k in divisor_list}
    def reduce(self, divisor_list):
        self.remainders = { k: self.remainders[k] % k for k in divisor_list}
    def eval(self, eval_str):
        self.remainders = {k:cmd_eval(old, eval_str) for k,old in self.remainders.items()}    
    def is_divisible_by(self, divisor):
        return True if self.remainders[divisor] == 0 else False


@dataclass
class Monkey:
    items: Deque[Union[Item, int]]
    op_str: str
    divisor: int
    true_tgt: int
    false_tgt: int
    inspect_ctr: int = 0
    
    def setup(self, divisor_list):
        for item in self.items:
            if isinstance(item, Item):
                item.setup(divisor_list)

    def inspect_and_reduce(self, divisor_list, item):
        item.eval(self.op_str)
        item.reduce(divisor_list)
        tgt = self.true_tgt if item.is_divisible_by(self.divisor) else self.false_tgt
        return tgt, item

    def inspect_and_divide(self, item, reduction=3):
        item = cmd_eval(item, self.op_str) // reduction
        tgt = self.true_tgt if item % self.divisor == 0 else self.false_tgt
        return tgt, item

    def inspect(self, divisor_list):
        self.inspect_ctr += 1
        item = self.items.popleft()
        if isinstance(item, int):
            return self.inspect_and_divide(item)          ### Part One
        else:
            return self.inspect_and_reduce(divisor_list, item)  ### Part Two


def simulate_round(monkey_dict, divisor_list):
    for _monkey_idx, monkey in monkey_dict.items():
        for item_idx in range(0, len(monkey.items)):
            tgt, val = monkey.inspect(divisor_list)
            monkey_dict[tgt].items.append(val)


def simulate_monkeys(monkey_dict, num_rounds):
    divisor_list = [monkey.divisor for _,monkey in monkey_dict.items()]
    for _, monkey in monkey_dict.items():
        monkey.setup(divisor_list)
    for _i in tqdm(range(0, num_rounds)):
        simulate_round(monkey_dict, divisor_list)


def study_monkeys(monkey_dict):
    activities = sorted([monkey.inspect_ctr for _,monkey in monkey_dict.items()])
    return activities[-1] * activities[-2]


def build_monkey(cmd_lines, part_one):
    m_id = int(cmd_lines[0][-3])
    item_list = [int(x) for x in cmd_lines[1].strip().split(':')[-1].split(',')]
    op_str = cmd_lines[2].strip().split('=')[-1]
    div = int(cmd_lines[3].strip().split(' ')[-1])
    tr_tgt = int(cmd_lines[4].strip().split(' ')[-1])
    f_tgt = int(cmd_lines[5].strip().split(' ')[-1])
    if part_one:
        return m_id, Monkey(deque(item_list), op_str, div, tr_tgt, f_tgt)
    else:
        return m_id, Monkey(deque([Item(x) for x in item_list]), op_str, div, tr_tgt, f_tgt)


def load_monkeys(filename: str, part_one=True):
    cmd_lines = []
    monkey_dict = {}
    with open(filename, 'r') as file:
        cmd_lines = file.readlines()    
    for line_idx in range(0, len(cmd_lines), 7):
        k, v = build_monkey( cmd_lines[line_idx:line_idx+7], part_one)
        monkey_dict[k] = v
    return monkey_dict
     

if __name__ == "__main__":
    time_start = perf_counter_ns()
    file_loc = r"C:\Users\Chris\Documents\CodeProjects\AdventOfCode\Yr2022\PuzzleData\puzzle_day_eleven.txt"
    
    part_one = False
    monkey_dict = load_monkeys(file_loc, part_one)
    if part_one:
        simulate_monkeys(monkey_dict, 20)
    else:
        simulate_monkeys(monkey_dict, 10000)
    monkey_business = study_monkeys(monkey_dict)
    
    print (f"Level of monkey business: {monkey_business}")       
    print(f"Elapsed time : {(perf_counter_ns() - time_start)/1000:.0f} microseconds")

File: Yr2022/src/day_eleven.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::time::Instant;
use std::collections::HashMap;



fn main() {   
    let start = Instant::now();
    let file_name= "C:\\Users\\Chris\\Documents\\CodeProjects\\AdventOfCode\\Yr2022\\PuzzleData\\puzzle_day_eleven.txt";
    let part_one = false;
    let monkey_business = load_monkeys(file_name, part_one);    
    let duration = start.elapsed();
    println!("Monkey Business: {}", monkey_business);
    println!("Time taken to execute: {} microseconds", duration.as_micros());    
}



pub struct Item  {
    val : u32,
    remainders: HashMap<u32, u32>,
}

impl Item {
    pub fn reduce(&mut self, divisor_list: &Vec<u32>) {
        for div in divisor_list {
            self.remainders.entry(*div).and_modify(|x| {*x = *x % div}).or_insert(self.val as u32);
            }
    }
    pub fn is_divisible_by(&mut self, div: u32) -> bool {
        return self.remainders[&div] == 0;
    }
}


pub struct Monkey {
    items: Vec<Item>,
    op_char: char,
    modifier: u32,
    divisor: u32,
    true_tgt: u32,
    false_tgt: u32,
    inspect_ctr: u32,    
}

impl Monkey {
    pub fn inspect(&mut self, divisor_list: &Vec<u32>, part_one: bool) -> (u32, Item) {
        let mut item = self.items.pop().expect("There should be an item");
        self.inspect_ctr += 1;
        match part_one {
        true => {  match self.op_char {
                '*' => {  item.val =  (item.val * self.modifier)/3 ; },
                '+' => {  item.val =  (item.val + self.modifier)/3 ; },
                _ => {  item.val =  (item.val * item.val)/3; },
            }  
            let tgt = if item.val % self.divisor  == 0 {self.true_tgt} else {self.false_tgt};
            return (tgt, item);
            },
        false => { match self.op_char {
                '*' => { for v in item.remainders.values_mut() {*v =  *v * self.modifier; }},
                '+' => { for v in item.remainders.values_mut() {*v =  *v + self.modifier; }},
                _ => { for v in item.remainders.values_mut() {*v =  *v * *v; }},
            }        
            item.reduce(divisor_list);
            let tgt = if item.is_divisible_by(self.divisor) {self.true_tgt} else {self.false_tgt};
            return (tgt, item);
        }    
        }
    }
    
    pub fn reduce_all(&mut self, divisor_list: &Vec<u32>) {
        for _idx in 0..self.items.len() {
            self.items[_idx].reduce(divisor_list);
        }
    }
}


fn simulate_round(monkey_map: &mut HashMap<u32, Monkey>, divisor_list: &Vec<u32>, part_one: bool) {
    for _monkey_idx_usize in 0..monkey_map.len() {
        let monkey_idx = _monkey_idx_usize as u32;
        while monkey_map[ &monkey_idx ].items.len() > 0 {
            let (tgt, item) = monkey_map.get_mut(&monkey_idx).map(|val| val.inspect(divisor_list, part_one)).expect("");
            monkey_map.entry(tgt).and_modify(|x| {x.items.push(item)});
        }
    }
}


fn simulate_monkeys(monkey_map: &mut HashMap<u32, Monkey>, num_rounds: usize, part_one:bool) {
    let divisor_list: Vec<u32> = monkey_map.iter().map(|(_k,v)| v.divisor ).collect();
    for monkey in monkey_map.values_mut() {
        monkey.reduce_all(&divisor_list);
    }
    for _i in 0..num_rounds {
        simulate_round(monkey_map, &divisor_list, part_one);
    }
}


fn get_monkey_business(monkey_map: &HashMap<u32, Monkey>) -> u32{
    let mut activities : Vec<u32> = monkey_map.values().map(|m| m.inspect_ctr).collect();
    activities.sort_by(|a,b| b.cmp(a) );    
    return activities[0] * activities[1];
}


fn load_monkeys(my_file_name: &str, part_one: bool) -> u32{    
    let mut monkey_map :HashMap<u32, Monkey> = HashMap::new();
    let file = File::open(my_file_name).expect("File does not exist");
    let buf_lines = BufReader::new(file).lines();
    let line_vec: Vec<String> = buf_lines.map(|l| l.expect("Fail to parse")).collect();

    for line_idx in (0..line_vec.len()).step_by(7) {
        let first_line: Vec<&str> = line_vec[line_idx].split(&[' ', ':']).collect();
        let m_id : u32 = first_line[1].parse().unwrap();

        let second_line : Vec<&str> = line_vec[line_idx + 1].split(&[':']).collect();
        let item_list_str: Vec<&str> = second_line[1].split(&[',']).collect();
        let item_list : Vec<u32> = item_list_str.iter().map(|s| s.trim().parse::<u32>().unwrap()).collect();

        let third_line : Vec<&str> = line_vec[line_idx + 2].split(&[' ']).collect();
        let modifier_str  = third_line[7];
        let mut arg_op_char = 'X';
        let mut arg_modifier = 0;
        match modifier_str {
            "old" => (),
            _ => {arg_op_char = third_line[6].chars().next().unwrap();
                 arg_modifier = modifier_str.parse::<u32>().unwrap();}
        }

        let fourth_line : Vec<&str> = line_vec[line_idx + 3].split(&[' ']).collect();
        let arg_divisor : u32 = fourth_line[5].parse::<u32>().unwrap();

        let fifth_line : Vec<&str> = line_vec[line_idx + 4].split(&[' ']).collect();
        let arg_t_tgt: u32 = fifth_line[9].parse::<u32>().unwrap();
        
        let sixth_line : Vec<&str> = line_vec[line_idx + 5].split(&[' ']).collect();
        let arg_f_tgt: u32 = sixth_line[9].parse::<u32>().unwrap();

        monkey_map.insert(m_id, Monkey {items:Vec::new(), op_char:arg_op_char, modifier:arg_modifier,
                                divisor:arg_divisor, true_tgt:arg_t_tgt, false_tgt:arg_f_tgt, inspect_ctr:0});
        for item_val in item_list {
            monkey_map.get_mut(&m_id).expect("").items.push(Item {val:item_val , remainders:HashMap::new()});
        }

    }
    
    let runs = if part_one {20} else {10000};
    simulate_monkeys(&mut monkey_map, runs, part_one);
    let monkey_business = get_monkey_business(&monkey_map);
    return monkey_business;
}

